src/Workspaces/Core/Portable/Editing/ImportAdderService.cs

```diff
@@ -56,6 +57,22 @@ public enum Strategy
                 _ => throw new InvalidEnumArgumentException(nameof(strategy), (int)strategy, typeof(Strategy)),
             };
 
+            if (importDirectivesToAdd.Count is 0)
+            {
+                return document.WithSyntaxRoot(root); //keep any added simplifier annotations
+            }
+
+            if (safe)
+            {
+                var annotation = new SyntaxAnnotation();
+                root = root.ReplaceNode(context, context.WithAdditionalAnnotations(annotation));
+                document = document.WithSyntaxRoot(root);
+                root = await document.GetSyntaxRootAsync().ConfigureAwait(false);
+                model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                root = MakeSafeToAddNamespaces(root, namespaceSymbols, model, document.Project.Solution.Workspace, cancellationToken);
```

Comment:
i don't really like waht i'm dseeing here because tehre are so many assignments to the same variable,a nd it's hard to tell what data is now valid.  for example, you ensure that you get an updated document with the root at one point, but then you let root diverge again.  same with your semantic model.

It feels quite dangerous to me, esp for the code lower down.